Add SendEvent to AccessGraphService

jakule opened the pull request:
Updated several proto files and `service.go` to support the addition of the SendEvent method in AccessGraphService. This method enables sending an event to the access graph service. Also updated the access graph watcher to include the method and send events as necessary. These changes provide the mechanisms needed for proper interaction with the access graph service.

I moved the generated protos to https://github.com/gravitational/teleport/pull/33286 

Note: This PR preserves the hacky logic of sending everything every time to TAG. This problem will be addressed in a separate PR.

jakule committed:
Add SendEvent to AccessGraphService

Updated several proto files and `service.go` to support the addition of the SendEvent method in AccessGraphService. This method enables sending an event to the access graph service. Also updated the access graph watcher to include the method and send events as necessary. These changes provide the mechanisms needed for proper interaction with the access graph service.

jakule committed:
Remove commented code in service.go

Cleaned up service.go by removing unused, commented-out code that pertains to the AccessGraphProcessor. This is part of a larger effort to clean up the codebase and improve readability. The code about reusing the connection and adding TLS are marked as TODOs, but they are not yet implemented or used, hence also have been removed. This also safeguards from potential bugs in the future due to confusion. Now the connection gets explicitly closed everytime.

jakule committed:
Add generated python proto

jakule committed:
Add generated python proto config

jakule committed:
Add missing pyi files

jakule committed:
Plugin for protocolbuffers/pyi was added to the buf-python.gen.yaml configuration file.

This addition ensures that we generate PYI files for static type checking in Python, a step that was previously missing. This helps our developers in catching potential type errors ahead of time, enhancing the overall code quality and reducing debugging time.

jakule committed:
Add new access graph service integration

This commit introduces a new access graph service to the Teleport API that synchronizes various resource types. It makes use of gRPC to safely and optimally connect with external services. With this feature, resources are being cached and monitored in real time. It increases the efficiency and speed of resource retrieval operations. Changes were also made in 'grpcserver.go' to allow component name to be used in the WatchEvents function, which aids in tracing operations. One thing to note is that tls support and ordering of sending resources are still missing and marked as TODO. These changes also required a modification in the 'service.go' to initialize and watch the access graph upon service start.

jakule committed:
Implement synchronization of access requests to the new AccessGraph service

Added a new function, `sendAccessRequests()`, to send all existing access requests to the AccessGraphService upon initialization. This is to ensure that the access requests are in sync on both the auth server and the new AccessGraph service. Also, `Namespace` parameter was added in the `listReq` in `getResourceEvents()` to specify the namespace of resources to retrieve, addressing missing details in the previous version. Removed `sendNodes()` function as resources should now be fetched via the new AccessGraph service.

jakule committed:
Remove generated proto

justinas commented on lib/service/service.go:
```diff
@@ -1860,6 +1878,35 @@ func (process *TeleportProcess) initAuthService() error {
+			// TODO(jakule): Very excessive retrying, but we need to make sure that
+			// the access graph is initialized before we start serving requests.
+			for {
+				if err := initializeAndWatchAccessGraph(process.ExitContext(), accessGraphAddr, authServer); err != nil {
```
Seems like this does not exit and instead wait forever when receiving SIGQUIT (`pkill -3 teleport`). Same issue as https://github.com/gravitational/teleport/pull/28356 , see that PR for the fix.

jakule replied:
Thanks. Wasn't aware.

jakule committed:
Add cache initialization status to SendEventResponse protobuf

Add a boolean field `cache_initialized` to the `SendEventResponse` message type in protobuf. This is to let the client know whether the cache was initialized successfully or not after an event is sent.

jakule committed:
Add SendResourceRequest and its sub-messages to the access graph.

The SendResourceRequest message has been added to allow sending resources to the access graph. Several sub-messages that are assignable to the SendResourceRequest's Resource field have also been added. These include structures representing different resource types such as User, Role, Server, and AccessRequest.

jakule committed:
GCI
